test_util/launch.py

```diff
@@ -1,19 +1,136 @@
 
+Commands:
+  create    reads the file given by --config-path, creates the cluster described therein
+            and finally dumps a JSON file to the path given in --info-path which can then
+            be used with the wait, describe, and delete calls
+  wait      Block until the cluster is up and running
+  describe  Return additional information about the composition of the cluster
+  delete    Destroying the provided cluster deployment
+
+Options:
+  --config-path=<path>  Path for config to create cluster from [default: config.yaml].
+  --info-path=<path>    JSON file output by create and consumed by wait, describe,
+                        and delete [default: cluster_info.json].
+"""
+import abc
+import json
+import os
+import sys
+import yaml
 from docopt import docopt
+from pkgpanda.util import load_json, load_string, write_json
+
+from test_util.aws import BotoWrapper, DcosCfSimple
+
+
+class AbstractLauncher(metaclass=abc.ABCMeta):
+    def create(self):
+        raise NotImplementedError()
+
+    def wait(self):
+        raise NotImplementedError()
+
+    def describe(self):
+        raise NotImplementedError()
+
+    def delete(self):
+        raise NotImplementedError()
+
+
+class AwsCloudformationLauncher(AbstractLauncher):
+    def __init__(self, config):
+        self.boto_wrapper = BotoWrapper(
+            config['region'], config['access_key_id'], config['secret_access_key'])
+        self.config = config
+
+    def create(self):
+        self.boto_wrapper.create_stack(
+            self.config['stack_name'], self.config['template_url'], self.config['parameters'])
+        return {
+            'type': 'cloudformation',
+            'stack_name': self.config['stack_name'],
+            'region': self.config['region'],
+            'access_key_id': self.config['access_key_id'],
+            'secret_access_key': self.config['secret_access_key']}
+
+    def wait(self):
+        # TODO: should this support the case where the cluster is being updated?
+        cf = self.get_stack()
+        status = cf.get_stack_details()['StackStatus']
+        if status == 'CREATE_IN_PROGRESS':
+            cf.wait_for_stack_creation(wait_for_poll_min=0)
+        elif status == 'CREATE_COMPLETE':
+            pass
+        else:
+            raise Exception('')
+
+    def describe(self):
+        cf = self.get_stack()
+        return {
+            'masters': convert_host_list(cf.get_master_ips()),
+            'private_agents': convert_host_list(cf.get_private_agent_ips()),
+            'public_agents': convert_host_list(cf.get_public_agent_ips())}
+
+    def delete(self):
+        self.get_stack().delete()
+
+    def get_stack(self):
+        """Returns the correct class interface depending how the AWS CF is configured
+        NOTE: only supports Simple Cloudformation currently
+        """
+        return DcosCfSimple(self.config['stack_name'], self.boto_wrapper)
+
+
+def get_launcher(config):
```

Comment:
this says it takes a config, but sometimes we pass it a cluster_info object which isn't the config.